### PR TITLE
buffer: use move construct to append/push_back/push_front

### PR DESCRIPTION
**Pull request by yuyuyu101**

### CODE REVIEW

File: src/common/buffer.cc [hunk 1]
```diff
@@ -1582,6 +1587,12 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
       push_back(bp);
   }
 
+  void buffer::list::append(ptr&& bp)
+  {
+    if (bp.length())
+      push_back(bp);
```
Comment (cbodley) [inline src/common/buffer.cc]: bp is an lvalue in the body of the function, so the call to push_back() needs to cast it again: `push_back(std::move(bp));`

Comment (yuyuyu101) [inline src/common/buffer.cc]: OMG, I correct my thought now....

File: src/include/buffer.h [hunk 2]
```diff
@@ -414,9 +415,14 @@ namespace buffer CEPH_BUFFER_API {
       _buffers.push_front(bp);
       _len += bp.length();
     }
+    void push_front(ptr&& bp) {
+      if (bp.length() == 0)
+	return;
+      _buffers.push_front(bp);
+      _len += bp.length();
```
Comment (cbodley) [inline src/include/buffer.h]: same, need to `std::move(bp)` when passing it to push_front(). then you'll need to call to bp.length() before it moves away:
```
_len += bp.length();
buffers.push_front(std::move(bp));
```

File: src/include/buffer.h [hunk 3]
```diff
@@ -424,9 +430,14 @@ namespace buffer CEPH_BUFFER_API {
       _buffers.push_back(bp);
       _len += bp.length();
     }
+    void push_back(ptr&& bp) {
+      if (bp.length() == 0)
+	return;
+      _buffers.push_back(bp);
+      _len += bp.length();
```
Comment (cbodley) [inline src/include/buffer.h]: see comment for push_front()

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -5473,10 +5469,8 @@ int BlueStore::_do_write_zero(
   uint64_t offset,
   uint64_t length)
 {
-  bufferptr z(length);
-  z.zero();
   bufferlist zl;
-  zl.push_back(z);
+  zl.appen_zero(length);
```
Comment (cbodley) [inline src/os/bluestore/BlueStore.cc]: typo? `append_zero`

**Comment by yuyuyu101** [comment]
done

File: src/include/buffer.h [hunk 5]
```diff
@@ -411,22 +412,32 @@ namespace buffer CEPH_BUFFER_API {
     void push_front(ptr& bp) {
       if (bp.length() == 0)
 	return;
+      _len += bp.length();
       _buffers.push_front(bp);
+    }
+    void push_front(ptr&& bp) {
+      if (bp.length() == 0)
+	return;
+      _buffers.push_front(std::move(bp));
```
Comment (cbodley) [inline src/include/buffer.h]: need to call bp.length() before push_front() here. it looks like you applied this fix to `void push_front(ptr& bp) {` instead

File: src/include/buffer.h [hunk 6]
```diff
@@ -411,22 +412,32 @@ namespace buffer CEPH_BUFFER_API {
...
     void push_back(const ptr& bp) {
       if (bp.length() == 0)
 	return;
+      _len += bp.length();
       _buffers.push_back(bp);
+    }
+    void push_back(ptr&& bp) {
+      if (bp.length() == 0)
+	return;
+      _buffers.push_back(std::move(bp));
```
Comment (cbodley) [inline src/include/buffer.h]: same here. you moved the call to bp.length() in `void push_back(const ptr& bp) {` instead

**Comment by yuyuyu101** [comment]
@cbodley thanks for quick comments, fixed

**Comment by cbodley** [comment]
@yuyuyu101 looks good to me :+1:

File: src/osdc/ObjectCacher.cc [hunk 7]
```diff
@@ -724,12 +724,10 @@ void ObjectCacher::bh_read_finish(int64_t poolid, sobject_t oid,
...
 
   if (r >= 0 && bl.length() < length) {
-    bufferptr bp(length - bl.length());
-    bp.zero();
     ldout(cct, 7) << "bh_read_finish " << oid << " padding " << start << "~"
-		  << length << " with " << bp.length() << " bytes of zeroes"
+		  << length << " with " << length - bl.length() << " bytes of zeroes"
 		  << dendl;
-    bl.push_back(bp);
+    bl.append_zero();
```
Comment (tchaikov) [inline src/osdc/ObjectCacher.cc]: should be
```c++
bl.append_zero(length - bl.length());
```

**Comment by tchaikov** [comment]
apart from https://github.com/ceph/ceph/pull/7455#discussion_r51439280, lgtm

might need to extract the change to bufferlist into another commit.

**Comment by yuyuyu101** [comment]
@tchaikov thanks, done.

**Comment by tchaikov** [comment]
lgtm with qa run.

**Comment by liewegas** [comment]
In file included from os/bluestore/KernelDevice.h:18:0,
                 from os/bluestore/KernelDevice.cc:22:
./os/fs/FS.h: In member function ‘void FS::aio_t::pread(uint64_t, uint64_t)’:
./os/fs/FS.h:79:32: error: ‘p’ was not declared in this scope
       io_prep_pread(&iocb, fd, p.c_str(), length, offset);
                                ^

**Comment by yuyuyu101** [comment]
@cbodley @tchaikov after tested locally, I found operator= is missing for rvalue, so I added this commit!

plz review again!

**Comment by cbodley** [comment]
@yuyuyu101 the operator= looks correct to me. i see that we have pretty good test coverage for the bufferlist constructors and operators in src/test/bufferlist.cc; could you add some test cases for the new operations?

**Comment by yuyuyu101** [comment]
@cbodley @liewegas updated

**Comment by cbodley** [comment]
@yuyuyu101 looks good, thanks!

**Comment by dillaman** [comment]
@yuyuyu101 Running ```git bisect```, it looks like ceph_test_librbd_fsx fails after its first clone validation due to commit dd4e6e1:

> truncating image image_client.0-clone1 from 0xea8df03 (overlap 0x1a7ae16) to 0x54bc6a
checking clone #0, image image_client.0 against file /home/jdillaman/archive/fsx-image_client.0-parent1
READ BAD DATA: offset = 0x0, size = 0x8abe78a, fname = image_client.0

The comparison fails on a section where there is no backing RBD object but the read operation returned a non-zeroed buffer.

Here is a simple reproducer: ceph_test_librbd_fsx -d -W -R -p 100 -P ~/archive -r 1 -w 1 -t 1 -h 1 -l 250000000 -S 4621 -N 1000 pool_client.0 image_client.0 -U

**Pull request merged**